crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1364,7 +1364,25 @@ impl<'db> TypeInferenceBuilder<'db> {
             simple: _,
         } = assignment;
 
-        let annotation_ty = self.infer_annotation_expression(annotation);
+        let mut annotation_ty = self.infer_annotation_expression(annotation);
+
+        // If the variable is annotation with SpecialForm then create a new class with name of the
+        // variable.
+        if let Type::Instance(class) = annotation_ty {
+            if class.is_known(self.db, KnownClass::SpecialForm) {
+                let target_name = target.as_name_expr().unwrap().id.clone();
```

Comment:
We should use another `if-let` here to guard against the possibility that the target isn't a simple Name (in which case we can just assume this isn't any known special form)